peer/peerwg/wireguard.go

```diff
@@ -0,0 +1,416 @@
+package peerwg
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+	"io"
+	"log"
+	"net"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/tabbed/pqtype"
+	"golang.org/x/xerrors"
+	"inet.af/netaddr"
+	"tailscale.com/ipn/ipnstate"
+	"tailscale.com/net/dns"
+	"tailscale.com/net/netns"
+	"tailscale.com/net/tsdial"
+	"tailscale.com/tailcfg"
+	"tailscale.com/types/ipproto"
+	"tailscale.com/types/key"
+	tslogger "tailscale.com/types/logger"
+	"tailscale.com/types/netmap"
+	"tailscale.com/wgengine"
+	"tailscale.com/wgengine/filter"
+	"tailscale.com/wgengine/magicsock"
+	"tailscale.com/wgengine/monitor"
+	"tailscale.com/wgengine/netstack"
+	"tailscale.com/wgengine/router"
+	"tailscale.com/wgengine/wgcfg/nmcfg"
+
+	"cdr.dev/slog"
+)
+
+func UUIDToInet(uid uuid.UUID) pqtype.Inet {
+	uid[0] = 0xfd
+	uid[1] = 0x7a
+	uid[2] = 0x11
+	uid[3] = 0x5c
+	uid[4] = 0xa1
+	uid[5] = 0xe0
+
+	return pqtype.Inet{
+		Valid: true,
+		IPNet: net.IPNet{
+			IP:   uid[:],
+			Mask: net.CIDRMask(128, 128),
+		},
+	}
+}
+
+func UUIDToNetaddr(uid uuid.UUID) netaddr.IP {
+	// fd7a:115c:a1e0
+	uid[0] = 0xfd
+	uid[1] = 0x7a
+	uid[2] = 0x11
+	uid[3] = 0x5c
+	uid[4] = 0xa1
+	uid[5] = 0xe0
+
+	return netaddr.IPFrom16(uid)
+}
+
+var logf tslogger.Logf = log.Printf
+
+type WireguardNetwork struct {
+	mu      sync.Mutex
+	logger  slog.Logger
+	Private key.NodePrivate
+	Disco   key.DiscoPublic
+
+	Engine   wgengine.Engine
+	Netstack *netstack.Impl
+	Magic    *magicsock.Conn
+
+	netMap    *netmap.NetworkMap
+	router    *router.Config
+	listeners map[listenKey]*listener
+}
+
+func NewWireguardNetwork(_ context.Context, logger slog.Logger, addrs []netaddr.IPPrefix) (*WireguardNetwork, error) {
+	var (
+		private      = key.NewNode()
+		public       = private.Public()
+		id, stableID = nodeIDs(public)
+	)
+
+	netMap := &netmap.NetworkMap{
+		NodeKey:    public,
+		PrivateKey: private,
+		Addresses:  addrs,
+		PacketFilter: []filter.Match{{
+			IPProto: []ipproto.Proto{ipproto.TCP, ipproto.UDP, ipproto.ICMPv4, ipproto.ICMPv6},
+			Srcs: []netaddr.IPPrefix{
+				netaddr.IPPrefixFrom(netaddr.IPv4(0, 0, 0, 0), 0),
+				netaddr.IPPrefixFrom(netaddr.IPv6Unspecified(), 0),
+			},
+			Dsts: []filter.NetPortRange{
+				{
+					Net: netaddr.IPPrefixFrom(netaddr.IPv4(0, 0, 0, 0), 0),
+					Ports: filter.PortRange{
+						First: 0,
+						Last:  65535,
+					},
+				},
+				{
+					Net: netaddr.IPPrefixFrom(netaddr.IPv6Unspecified(), 0),
+					Ports: filter.PortRange{
+						First: 0,
+						Last:  65535,
+					},
+				},
+			},
+			Caps: []filter.CapMatch{},
+		}},
+	}
+	netMap.SelfNode = &tailcfg.Node{
+		ID:         id,
+		StableID:   stableID,
+		Key:        public,
+		Addresses:  netMap.Addresses,
+		AllowedIPs: append(netMap.Addresses, netaddr.MustParseIPPrefix("::/0")),
+		Endpoints:  []string{},
+		DERP:       DefaultDerpHome,
+	}
+
+	linkMon, err := monitor.New(logf)
+	if err != nil {
+		return nil, xerrors.Errorf("create link monitor: %w", err)
+	}
```

Comment:
Some comments on what these things do would be helpful. It's a bit difficult for me to understand the chaining.